runtime/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSubscribeFactory.java

```diff
@@ -575,11 +599,17 @@ private void onMqttWindow(
             {
                 retained.doKafkaWindow(traceId, authorization, budgetId, padding, capabilities);
             }
-            else if (messages.messageSlotOffset != messages.messageSlotLimit)
+            else
             {
-                messages.flushData(traceId, authorization, budgetId);
+                messages.values().forEach(m ->
+                {
+                    if (m.messageSlotOffset != m.messageSlotLimit)
+                    {
+                        m.flushData(traceId, authorization, budgetId);
+                    }
+                });
```

Comment:
Test this condition inside a method on `KafkaMessagesProxy` instead so this becomes a simpler one line lambda.